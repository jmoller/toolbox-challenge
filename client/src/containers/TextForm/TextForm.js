import React, { Component } from "react";
import { Field } from "redux-form";
import {
  FormGroup,
  FormControl,
  Button,
} from "react-bootstrap";

// Decoupled Message Input component for testing purposes
// Desc: This is the input component used by Field in redux-form
export const MessageInput = ({
  input,
  meta: { touched, error },
  ...custom
}) => {
  const hasError = touched && error !== undefined;

  const controlClass = hasError ? "message has-error" : "message";
  return (
    <React.Fragment>
      <FormControl
        type="text"
        placeholder="Example: Hello World!"
        className={controlClass}
        {...input}
        {...custom}
      />

      {hasError && (
        <p
          style={{
            marginTop: "1rem",
            color: "#c10000",
            letterSpacing: ".1rem",
            fontWeight: "600"
          }}
        >
          {error}
        </p>
      )}
    </React.Fragment>
  );
};

// Unconnected Textform component that will be decorated by the TextFormContainer 
// By spliting up the Form into two components, we can safly test TextForm as a regular component
class TextForm extends Component {
  submit(values, dispatch) {
    this.props.postTexts(values);
  }

  render() {
    const { handleSubmit } = this.props;
    const formStyle = {
      display: "flex",
      flexDirection: "column",
      justifyContent: "space-around",
      minHeight: "30vh",
      width: "25rem"
    };

    return (
      <form style={formStyle} onSubmit={handleSubmit(this.submit.bind(this))}>
        <FormGroup>
          <Field name="message" component={MessageInput} />
        </FormGroup>

        <Button style={{ width: "100%" }} bsStyle="primary" type="submit">
          Send Message
        </Button>
      </form>
    );
  }
}

export default TextForm;
