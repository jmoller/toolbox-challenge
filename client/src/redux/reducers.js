import TextsReducer from "./texts/reducer";
import { reducer as formReducer } from "redux-form";

export default {
  TextsReducer,
  form: formReducer
};
