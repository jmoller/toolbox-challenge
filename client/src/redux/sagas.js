import { all } from "redux-saga/effects";
import textsSagas from "./texts/saga";

export default function* rootSaga(getState) {
  yield all([textsSagas()]);
}
