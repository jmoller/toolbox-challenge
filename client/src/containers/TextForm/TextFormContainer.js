import { connect } from "react-redux";
import TextForm from "./TextForm";
import { reduxForm } from "redux-form";
import { postTexts } from "../../redux/texts/actions";
import {validateMessageInput} from '../../utilities/validation';

// The decorated Form component
// Desc: It takes the TextForm as a parameter in the reduxForm function to return a decorated container
// that let us make use of reduxForm functionalities
let TextFormContainer = reduxForm({
  form: "messageForm",
  validate: validateMessageInput,
  fields: ["message"]
})(TextForm);

TextFormContainer = connect(
  null,
  { postTexts }
)(TextFormContainer);

export default TextFormContainer;
