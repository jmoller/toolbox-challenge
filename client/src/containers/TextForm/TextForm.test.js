import React from 'react';
import {shallow, mount} from 'enzyme';


import {MessageInput} from './TextForm';
import TextForm from './TextForm';
import TexFormContainer from './TextFormContainer';

import { reducer as formReducer } from 'redux-form';
import { reduxForm } from 'redux-form';
import { createStore, combineReducers } from 'redux';
import { Provider } from 'react-redux';
import TextFormContainer from './TextFormContainer';

describe('MessageInput', () => {
  describe('is empty', () => {
    const props = {
      input: {},
      meta: { touched: true, error: 'A message is required' },
    }
    const messageInput = shallow(<MessageInput {...props}/>);

    it('displays an error message', () => {
      expect(messageInput.find('p').text()).toEqual(props.meta.error);
    });
  });

  describe('has text', () => {
    const props = {
      input: {},
      meta: { touched: true },
    }
    const messageInput = shallow(<MessageInput {...props}/>);

    it('does not display a paragraph', () => {
      expect(messageInput.find('p').length).toEqual(0);
    });
  });

});


describe('TextForm', () => {
  const mockFn = jest.fn();
  const props = {
    fields: {
      message: {
        value: '',
        touched: false,
        error: null
      }
    },
    values: {
      message: ''
    },
    handleSubmit: fn => fn,
    postTexts: mockFn
  } 
  

  const textForm = shallow(<TextForm {...props}/>)

  it('renders properly on screen', () => {
    expect(textForm).toMatchSnapshot();
  });

  describe('the form gets submited by the user', () => {
    beforeEach(() => {
      textForm.find('form').simulate('submit');
    });

    it('executes the callback function', () => {
      expect(mockFn).toHaveBeenCalledTimes(1);
      
    });
  });

});

describe('TextFormContainer', () => {
  let store = createStore(combineReducers({ form: formReducer }));
  
  const tfContainer = mount(
    <Provider store={store}>
      <TextFormContainer />
    </Provider>)
  

   tfContainer.find('.btn-primary').simulate('click');
   

} );
