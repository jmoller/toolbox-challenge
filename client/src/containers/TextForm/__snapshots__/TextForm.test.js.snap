// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextForm renders properly on screen 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextForm
    fields={
        Object {
            "message": Object {
              "error": null,
              "touched": false,
              "value": "",
            },
          }
    }
    handleSubmit={[Function]}
    postTexts={[Function]}
    values={
        Object {
            "message": "",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <FormGroup
          bsClass="form-group"
>
          <Field
                    component={[Function]}
                    name="message"
          />
</FormGroup>,
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="primary"
          disabled={false}
          style={
                    Object {
                              "width": "100%",
                            }
          }
          type="submit"
>
          Send Message
</Button>,
      ],
      "onSubmit": [Function],
      "style": Object {
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-around",
        "minHeight": "30vh",
        "width": "25rem",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "form-group",
          "children": <Field
            component={[Function]}
            name="message"
/>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "name": "message",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": false,
          "block": false,
          "bsClass": "btn",
          "bsStyle": "primary",
          "children": "Send Message",
          "disabled": false,
          "style": Object {
            "width": "100%",
          },
          "type": "submit",
        },
        "ref": null,
        "rendered": "Send Message",
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <FormGroup
            bsClass="form-group"
>
            <Field
                        component={[Function]}
                        name="message"
            />
</FormGroup>,
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="primary"
            disabled={false}
            style={
                        Object {
                                    "width": "100%",
                                  }
            }
            type="submit"
>
            Send Message
</Button>,
        ],
        "onSubmit": [Function],
        "style": Object {
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "space-around",
          "minHeight": "30vh",
          "width": "25rem",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": <Field
              component={[Function]}
              name="message"
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "name": "message",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": false,
            "bsClass": "btn",
            "bsStyle": "primary",
            "children": "Send Message",
            "disabled": false,
            "style": Object {
              "width": "100%",
            },
            "type": "submit",
          },
          "ref": null,
          "rendered": "Send Message",
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
